internal/topo/node/window_inc_agg_op.go

```diff
@@ -180,8 +181,293 @@ func (co *CountWindowIncAggOp) emit(ctx api.StreamContext, errCh chan<- error) {
+	currWindow *IncAggWindow
+}
+
+func NewTumblingWindowIncAggOp(o *WindowIncAggOperator) *TumblingWindowIncAggOp {
+	op := &TumblingWindowIncAggOp{
+		WindowIncAggOperator: o,
+		Interval:             o.windowConfig.Interval,
+	}
+	return op
+}
+
+func (to *TumblingWindowIncAggOp) exec(ctx api.StreamContext, errCh chan<- error) {
+	to.ticker = timex.GetTicker(to.Interval)
+	defer func() {
+		to.ticker.Stop()
+	}()
+	fv, _ := xsql.NewFunctionValuersForOp(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case input := <-to.input:
+			now := timex.GetNow()
+			data, processed := to.commonIngest(ctx, input)
+			if processed {
+				continue
+			}
+			switch row := data.(type) {
+			case *xsql.Tuple:
+				if to.currWindow == nil {
+					to.currWindow = newIncAggWindow(ctx, now)
+				}
+				name := calDimension(fv, to.Dimensions, row)
+				incAggCal(ctx, name, row, to.currWindow, to.aggFields)
+			}
+		case <-to.ticker.C:
+			if to.currWindow != nil {
+				to.emit(ctx, errCh)
+			}
+		}
+	}
+}
+
+func (to *TumblingWindowIncAggOp) emit(ctx api.StreamContext, errCh chan<- error) {
+	results := &xsql.WindowTuples{
+		Content: make([]xsql.Row, 0),
+	}
+	for _, incAggRange := range to.currWindow.DimensionsIncAggRange {
+		for name, value := range incAggRange.fields {
+			incAggRange.lastRow.Set(name, value)
+		}
+		results.Content = append(results.Content, incAggRange.lastRow)
+	}
+	results.WindowRange = xsql.NewWindowRange(to.currWindow.StartTime.UnixMilli(), time.Now().UnixMilli())
+	to.currWindow = nil
+	to.Broadcast(results)
+}
+
+type SlidingWindowIncAggOp struct {
+	*WindowIncAggOperator
+	triggerCondition ast.Expr
+	Length           time.Duration
+	Delay            time.Duration
+	delayWindowList  []*IncAggWindow
+	currWindowList   []*IncAggWindow
+	taskCh           chan *IncAggOpTask
+}
+
+type IncAggOpTask struct {
+	window *IncAggWindow
+}
+
+func NewSlidingWindowIncAggOp(o *WindowIncAggOperator) *SlidingWindowIncAggOp {
+	op := &SlidingWindowIncAggOp{
+		WindowIncAggOperator: o,
+		triggerCondition:     o.windowConfig.TriggerCondition,
+		Length:               o.windowConfig.Length,
+		Delay:                o.windowConfig.Delay,
+		currWindowList:       make([]*IncAggWindow, 0),
+		delayWindowList:      make([]*IncAggWindow, 0),
+		taskCh:               make(chan *IncAggOpTask, 1024),
+	}
+	return op
+}
+
+func (so *SlidingWindowIncAggOp) exec(ctx api.StreamContext, errCh chan<- error) {
+	fv, _ := xsql.NewFunctionValuersForOp(ctx)
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case input := <-so.input:
+			now := timex.GetNow()
+			data, processed := so.commonIngest(ctx, input)
+			if processed {
+				continue
+			}
+			switch row := data.(type) {
+			case *xsql.Tuple:
+				so.gcIncAggWindow(now)
+				if so.Delay > 0 {
+					so.appendDelayIncAggWindow(ctx, errCh, fv, row, now)
+					continue
+				}
+				so.appendIncAggWindow(ctx, errCh, fv, row, now)
+				if len(so.currWindowList) > 0 && so.isMatchCondition(ctx, fv, row) {
```

Comment:
isMatchCondition calculate mutliple times? Should this put into if so.Delay <=0?